### PR TITLE
Created a simple structure aware fuzzer

### PR DESCRIPTION
**Pull request by zd-project**
This PR creates a simple structure aware fuzzer that can generate arithmetic operations with simple literals. Operands are not necessarily numerical values.

### CODE REVIEW

File: zetasql/fuzzing/oss_fuzz.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    return false;
+  }
+  fs::path tzdataDir = originDir / "data/zoneinfo/";
+  if (setenv("TZDIR", tzdataDir.c_str(), OVERWRITE)) {
+    return false;
+  }
+  return true;
+}
+
+}
```
Comment (evmaus) [inline zetasql/fuzzing/oss_fuzz.h]: Nit:  End namespace with a comment // $(namespace).

File: zetasql/fuzzing/zetasql_expression_fuzzer.cc [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+#include <iostream>
+
+#include "libprotobuf_mutator/src/libfuzzer/libfuzzer_macro.h"
+#include "zetasql/public/evaluator.h"
+#include "zetasql/fuzzing/oss_fuzz.h"
+#include "zetasql/fuzzing/zetasql_expression_grammar.pb.h"
+#include "zetasql/fuzzing/zetasql_expression_proto_to_string.h"
+
+using namespace zetasql_expression_grammar;
```
Comment (evmaus) [inline zetasql/fuzzing/zetasql_expression_fuzzer.cc]: Per the Google style guide, using namespace statements should be avoided.

Instead, consider using specific class names (e.x. using zetasql_expression_grammer::Expression).

File: zetasql/fuzzing/zetasql_expression_grammar.proto [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+// TODO: Support high precision value representation
+message NumericLiteral {
+    required bytes value = 1;
+}
+
+message CompoundExpr {
+    oneof compound_oneof {
+        AdditiveOperation additive = 1;
```
Comment (evmaus) [inline zetasql/fuzzing/zetasql_expression_grammar.proto]: Out of curiousity-- is there a reason we're separating addition/subtraction and multiplication/division?  Since they're all binary operations, this is a little odd.

File: zetasql/fuzzing/zetasql_expression_proto_to_string.cc [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+CONV_FN(LiteralExpr, literal) {
+  using LitExprType = LiteralExpr::LiteralOneofCase;
+  std::string ret;
+  switch (literal.literal_oneof_case()) {
+    case LitExprType::kSpecialLiteral:
+      switch (literal.special_literal()) {
+        case LiteralExpr::V_NULL:
+          return "NULL";
+        default:
+          std::abort();
```
Comment (evmaus) [inline zetasql/fuzzing/zetasql_expression_proto_to_string.cc]: Consider outputting a logging statement in addition to just crashing here.

File: zetasql/fuzzing/zetasql_expression_proto_to_string.cc [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+      return IntegerLiteralToString(literal.integer_literal());
+    case LitExprType::kNumericLiteral:
+      return NumericLiteralToString(literal.numeric_literal());
+    default:
+      return HandleDefault();
+  }
+}
+
+CONV_FN(IntegerLiteral, integer) {
+  using IntergerType = IntegerLiteral::IntegerOneofCase;
```
Comment (evmaus) [inline zetasql/fuzzing/zetasql_expression_proto_to_string.cc:95]: Nit:  IntegerType*

File: zetasql/fuzzing/zetasql_expression_proto_to_string.h [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+// limitations under the License.
+//
+
+#ifndef ZETASQL_FUZZING_ZETASQL_EXPRESSION_PROTO_TO_STRING_H_
+#define ZETASQL_FUZZING_ZETASQL_EXPRESSION_PROTO_TO_STRING_H_
+
+#include "zetasql/fuzzing/common.h"
+#include "zetasql/fuzzing/zetasql_expression_grammar.pb.h"
+
+using namespace zetasql_expression_grammar;
```
Comment (evmaus) [inline zetasql/fuzzing/zetasql_expression_proto_to_string.h]: Same comment about "using namespace".

File: zetasql/fuzzing/zetasql_expression_proto_to_string_test.cc [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+//
+
+#include <memory>
+#include <string>
+
+#include "zetasql/fuzzing/zetasql_expression_grammar.pb.h"
+#include "zetasql/fuzzing/zetasql_expression_proto_to_string.h"
+#include "gtest/gtest.h"
+
+using namespace zetasql_expression_grammar;
```
Comment (evmaus) [inline zetasql/fuzzing/zetasql_expression_proto_to_string_test.cc]: Same comment about using namespace.

**Review by evmaus — CHANGES_REQUESTED** [review]
Few changes, but nothing major.

Tests seem pretty great!

File: zetasql/fuzzing/zetasql_expression_proto_to_string.cc [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    case ExprType::kExpr:
+      return CompoundExprToString(expr.expr());
+    default:
+      return HandleDefault();
+  }
+}
+
+CONV_FN(LiteralExpr, literal) {
+  using LitExprType = LiteralExpr::LiteralOneofCase;
+  std::string ret;
```
Comment (shumway) [inline zetasql/fuzzing/zetasql_expression_proto_to_string.cc]: Why is this here?

File: zetasql/fuzzing/zetasql_expression_proto_to_string.cc [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    case CompoundExprType::kMultiplicative:
+      return MultiplicativeOperationToString(comp_expr.multiplicative());
+    default:
+      return HandleDefault();
+  }
+};
+
+CONV_FN(AdditiveOperation, additive) {
+  std::string op;
+  if (additive.operator_() == AdditiveOperation::PLUS) {
```
Comment (shumway) [inline zetasql/fuzzing/zetasql_expression_proto_to_string.cc]: For consistency, just make this a switch, too. Then you don't need the comment and you put PLUS and MINUS on equal footing, and handle unexpected values in default.

The same thing for the next function with MULTIPLY  and DIVIDE.

File: zetasql/fuzzing/zetasql_expression_proto_to_string.cc [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    op = " - ";
+  }
+  return ExpressionToString(additive.lhs()) + op + ExpressionToString(additive.rhs());
+}
+
+CONV_FN(MultiplicativeOperation, multiplicative) {
+  std::string op;
+  if (multiplicative.operator_() == MultiplicativeOperation::MULTIPLY) {
+    op = " * ";
+  } else { // if multiplicative.operator_() == AdditiveOperation::DIVIDE
```
Comment (shumway) [inline zetasql/fuzzing/zetasql_expression_proto_to_string.cc]: MulitplicativeOperation

File: zetasql/fuzzing/zetasql_expression_proto_to_string_test.cc [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+
+namespace zetasql_fuzzer {
+
+TEST(ProtoToStringTest, UninitializedOneOfExprTest) {
+  Expression expr;
+  EXPECT_STREQ(ExpressionToString(expr).c_str(), "0");
```
Comment (shumway) [inline zetasql/fuzzing/zetasql_expression_proto_to_string_test.cc]: Google Test works fine with stings. No reason to print in char* pointers and EXPECT_STREQ Usual syntax is:

EXPECT_THAT(ExpressionToString(expr), Eq("0"))

with "using testing::Eq;" above.

Comment (shumway) [reply zetasql/fuzzing/zetasql_expression_proto_to_string_test.cc]: Or

EXPECT_EQ(ExpressionToString(expr), "0");

File: zetasql/fuzzing/zetasql_expression_proto_to_string_test.cc [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+#include <memory>
+#include <string>
+
+#include "zetasql/fuzzing/zetasql_expression_grammar.pb.h"
+#include "zetasql/fuzzing/zetasql_expression_proto_to_string.h"
+#include "gtest/gtest.h"
+
+using namespace zetasql_expression_grammar;
+
+namespace zetasql_fuzzer {
```
Comment (shumway) [inline zetasql/fuzzing/zetasql_expression_proto_to_string_test.cc:26]: Customary to add an anonymous namespace  inside to make sure that test code is never linked as a library.

namespace zetasql_fuzzer {
namespace {

// Test code and any helper functions or other code here.

} // namespace
} // namespace zetasql_fuzzer

File: zetasql/fuzzing/zetasql_expression_proto_to_string.h [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace zetasql_fuzzer {
+
+CONV_FN(Expression, expr);
+
+} //name space zetasql_fuzzer
```
Comment (shumway) [inline zetasql/fuzzing/zetasql_expression_proto_to_string.h]: nit:  Two spaces before // and one space after and namespace, not name space

}  // namespace zetasql_fuzzer

File: zetasql/fuzzing/common.h [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Common utilities for fuzzers
+
+#ifndef ZETASQL_FUZZING_COMMON_H_
+#define ZETASQL_FUZZING_COMMON_H_
+
+#define CONV_FN(TYPE, VAR_NAME) std::string TYPE##ToString(const TYPE& VAR_NAME)
```
Comment (shumway) [inline zetasql/fuzzing/common.h]: Why not just pure C++? It's a nice language.

File: zetasql/fuzzing/zetasql_expression_proto_to_string.cc [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+
+namespace zetasql_fuzzer {
+
+std::string HandleDefault() {
+  return "0";
+}
+
+// Forward declaration
+CONV_FN(Expression, expr);
+
```
Comment (shumway) [inline zetasql/fuzzing/zetasql_expression_proto_to_string.cc:51]: std::string ExpressionToString(const Expression& expr);

Losing the macro is much better in my opinion. It makes all design choices explicit, and it's more likely to be understood by IDEs and other code tools (including people who use vi /s and grep). Otherwise I have to keep checking what the macro does and making sure no one changed it to add side effects.

**Review by shumway — CHANGES_REQUESTED** [review]
This looks great!

I'd see CONV_FN removed (or defended) and the EXPECT_STREQ removed before approving.

File: zetasql/fuzzing/zetasql_expression_fuzzer.cc (hunk 2 above)
Comment (zd-project) [reply zetasql/fuzzing/zetasql_expression_fuzzer.cc]: So the convention was assumed from SQLite: https://chromium.googlesource.com/chromium/src/third_party/+/refs/heads/master/sqlite/fuzz/sql_query_proto_to_string.cc

It seems that we need to primarily/extensively use functions generated by protobuf. In an ever growing file that depends on various declared protobuf messages, declaring "using" for each class may unnecessarily bloat the codebase and wouldn't help much in readability.

File: zetasql/fuzzing/zetasql_expression_grammar.proto (hunk 3 above)
Comment (zd-project) [reply zetasql/fuzzing/zetasql_expression_grammar.proto]: I guess we can do that too. I was completely mirroring the definition used by the Bison parser

File: zetasql/fuzzing/zetasql_expression_proto_to_string.cc (hunk 4 above)
Comment (zd-project) [reply zetasql/fuzzing/zetasql_expression_proto_to_string.cc]: Would you suggest using `assert(false && "message")`;

File: zetasql/fuzzing/common.h (hunk 14 above)
Comment (zd-project) [reply zetasql/fuzzing/common.h]: So the convention is assume here: https://chromium.googlesource.com/chromium/src/third_party/+/refs/heads/master/sqlite/fuzz/sql_query_proto_to_string.cc#22

I think using macro does clean up code a little. If we need to cover the complete the entire syntax structure, we'll need **many** #ToString methods, as demonstrated in the code above. Having a macro here minimizes the probability of getting a typo. I do agree however, that CONV_FN may be a bad name for revealing its intention.

File: zetasql/fuzzing/zetasql_expression_proto_to_string.cc (hunk 15 above)
Comment (zd-project) [reply zetasql/fuzzing/zetasql_expression_proto_to_string.cc:51]: We can pull back the CONV_FN macro back into the cc file. In this way people don't need to go to "common.h" to understand what it does. In the "to_string.h" we can write out the complete function name "ExpressionToString" as the public API is more likely to be studied. Again, CONV_FN seems to be just as a convenience shorthand since we expect to use it a lot.

File: zetasql/fuzzing/zetasql_expression_proto_to_string.cc (hunk 8 above)
Comment (zd-project) [reply zetasql/fuzzing/zetasql_expression_proto_to_string.cc]: I believe it's debug residual that I forgot to remove.

File: zetasql/fuzzing/zetasql_expression_proto_to_string.cc (hunk 4 above)
Comment (evmaus) [reply zetasql/fuzzing/zetasql_expression_proto_to_string.cc]: So there is probably a CHECK macro available through abseil or somewhere else in zetasql.  I'd look for that first.

Comment (shumway) [reply zetasql/fuzzing/zetasql_expression_proto_to_string.cc]: Be careful about asserts, since they sometimes get removed in production.

I usually use LOG(FATAL) << "message"; but I don't know if you have access to that logging here.

File: zetasql/fuzzing/zetasql_expression_proto_to_string.cc (hunk 15 above)
Comment (shumway) [reply zetasql/fuzzing/zetasql_expression_proto_to_string.cc:51]: OK, it's a matter of style and your code. :)

File: zetasql/fuzzing/zetasql_expression_fuzzer.cc (hunk 2 above)
Comment (evmaus) [reply zetasql/fuzzing/zetasql_expression_fuzzer.cc]: I believe the sqlite code is actually part of the sqlite project.  They probably have a different (potentially conflicting!) style guide.  :)

So the current guidance is here:  https://google.github.io/styleguide/cppguide.html#Namespaces and explicitly forbids "using namespace foo" as it pollutes the namespace.  This is generally a maintainability concern--if you have a long file and one "using namespace ..." and you hit a name collision, you have a LOT of changes/cleanup/etc. to do to fix the problem.  (This gets even worse in headers.)

Frankly, this isn't a rule I like all that much for .cc files (I think it makes a lot of sense for headers).  You will often see a whole slew of "using namespace1::namespace2::class1; using namespace1::namespace2::class2; ..." in various Google code because of it, though, and I'd lean towards changing this, simple on the "when in Rome..." philosophy.  :)

File: zetasql/fuzzing/zetasql_expression_grammar.proto (hunk 3 above)
Comment (evmaus) [reply zetasql/fuzzing/zetasql_expression_grammar.proto]: So the reason they'd be separated in the Bison parser is to allow you to easily preserve order of operations in the parse tree.  Since we don't need to do that, it's probably cleaner to consolidate them.  :)

File: zetasql/fuzzing/common.h (hunk 14 above)
Comment (evmaus) [reply zetasql/fuzzing/common.h]: I also agree with Zhengdai here--I think the macro actually does clean things up slightly.

I do think it'd make sense to move it into the file it's used in (partly just to keep it scoped as close to the usage as possible and prevent it from spreading without us thinking more about it).  I'd also like to see some comments that show the usage & effect. :)

File: zetasql/fuzzing/zetasql_expression_proto_to_string.cc (hunk 4 above)
Comment (zd-project) [reply zetasql/fuzzing/zetasql_expression_proto_to_string.cc]: @shumway so likely we don't have logging function in OSS-Fuzz as it's not provided in the interface. Currently fuuzers are *always* compiled compiled with Bazel debug mode on so I'm not too worried about it. But I agree that maybe some kind of macro may help.

File: zetasql/fuzzing/zetasql_expression_proto_to_string.cc [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+  std::cerr << error << std::endl;
+  std::abort();
+}
+
+inline std::string HandleDefault() {
+  return "0";
+}
+
+inline std::string Padded(const std::string& s) {
+  return " " + s + " ";
```
Comment (evmaus) [inline zetasql/fuzzing/zetasql_expression_proto_to_string.cc]: Nit:  Might be very slightly faster to use absl::StrCat here.

File: zetasql/fuzzing/zetasql_expression_proto_to_string.cc [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+  case BinaryOperation::DIVIDE:
+    return "/";
+  default:
+    HandleUndefined("Undefined Binary Operator");
+  }
+}
+
+TO_STRING(BinaryOperation, binary_operation) {
+  std::string operator_str = GetBinaryOperator(binary_operation.op());
+  return ExpressionToString(binary_operation.lhs()) +
```
Comment (evmaus) [inline zetasql/fuzzing/zetasql_expression_proto_to_string.cc]: Not sure why I didn't catch this before--consider using absl::StrCat to make this a bit more efficient.

**Review by evmaus — COMMENTED** [review]
LGTM aside from two opportunities to use absl::StrCat instead of slower C++ string concatenation.

File: zetasql/fuzzing/zetasql_expression_proto_to_string.cc (hunk 16 above)
Comment (zd-project) [reply zetasql/fuzzing/zetasql_expression_proto_to_string.cc]: This is a primitive interface to be improved upon later. We'll likely include variable-length white-space characters in between.

**Review by evmaus — APPROVED** [review]
LGTM.

**Review by shumway — APPROVED** [review]
Looks great!

**Review by evmaus — APPROVED** [review]

**Review by shumway — APPROVED** [review]
